mmdet/core/hook/wandblogger_hook.py

```diff
@@ -0,0 +1,466 @@
+        metadata = dict(epoch=runner.epoch + 1, **eval_results)
+        return metadata
+
+    def _is_best_ckpt(self, metadata):
+        """Check if the current checkpoint has the best metric score.
+
+        Args:
+            metadata (dict): Metadata associated with the checkpoint.
+
+        Returns:
+            bool: Returns True, if the checkpoint has the best metric score.
+        """
+        keys = list(metadata.keys())
+        map_metrics = [key for key in keys if 'mAP' in key]
+        ar_metrics = [key for key in keys if 'AR' in key]
+        if len(map_metrics) > 0:
+            map_score = metadata.get(map_metrics[0], None)
+            return self._is_best_score(map_score)
+        elif len(ar_metrics) > 0:
+            ar_score = metadata.get(ar_metrics[0], None)
+            return self._is_best_score(ar_score)
+        else:
+            return False
+
+    def _is_best_score(self, score):
+        if score is None:
+            return
+
+        if score > self.best_score:
+            self.best_score = score
+            return True
+        else:
+            return False
+
+    def _init_data_table(self):
+        """Initialize the W&B Tables for validation data."""
+        columns = ['image_name', 'image']
+        self.data_table = self.wandb.Table(columns=columns)
+
+    def _init_pred_table(self):
+        """Initialize the W&B Tables for model evaluation."""
+        columns = ['epoch', 'image_name', 'ground_truth', 'prediction'] + list(
+            self.class_id_to_label.values())
+        self.eval_table = self.wandb.Table(columns=columns)
+
+    def _add_ground_truth(self):
+        num_total_images = len(self.val_dataset)
+        if self.num_eval_images > num_total_images:
+            warnings.warn(
+                'The num_eval_images is greater than the total number '
+                'of validation samples. The complete validation set '
+                'will be logged.', UserWarning)
+        self.num_eval_images = min(self.num_eval_images, num_total_images)
+
+        classes = self.val_dataset.get_classes()
+        self.class_id_to_label = {id: name for id, name in enumerate(classes)}
+        self.class_set = self.wandb.Classes([{
+            'id': id,
+            'name': name
+        } for id, name in self.class_id_to_label.items()])
+
+        for idx in range(self.num_eval_images):
+            data_sample = self.val_dataset.prepare_test_img(idx)
+            data_ann = self.val_dataset.get_ann_info(idx)
+
+            img_meta = data_sample['img_metas'][0].data
+            image = data_sample['img'][0].data
+            bboxes = data_ann['bboxes']
+            labels = data_ann['labels']
+            scale_factor = img_meta['scale_factor']
+
+            image_name = img_meta['ori_filename']
+            # permute to get channel last configuration, followed by bgr -> rgb
+            image = image.permute(
+                1, 2, 0).numpy()[..., ::-1]  # Not true for every dataloader
```

Comment:
What is the best way to know if the loaded image is in BGR or RGB mode?

One of my local testing scripts is loading the images in BGR format, thus I am doing this fancy slicing to convert to RGB. When I use `tools/train.py` with a tiny coco dataset, the images are loaded in RGB format.

What is the default loading mode for images (BGR or RGB)? Is there a way to programmatically know the mode in which it's loaded?

Reply:
You can use the first pipline `LoadImageFromFile` in the val dataset to get the image. Now, the `prepare_test_img` uses entire pipelines, which is not necessary.